File: src/builtin/time.rs
use crate::value::Value;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Mutex, OnceLock};
use std::thread;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

static TIMER_SEQ: AtomicI64 = AtomicI64::new(1);
static TIMERS: OnceLock<Mutex<HashMap<i64, Instant>>> = OnceLock::new();
static BOOT_INSTANT: OnceLock<Instant> = OnceLock::new();

fn timers() -> &'static Mutex<HashMap<i64, Instant>> {
    TIMERS.get_or_init(|| Mutex::new(HashMap::new()))
}

pub fn create_time_module() -> Value {
    let mut ns = HashMap::new();
    insert_fn(&mut ns, "now", time_now);
    insert_fn(&mut ns, "hrtimeMs", time_hrtime_ms);
    insert_fn(&mut ns, "sleep", time_sleep);
    insert_fn(&mut ns, "startTimer", time_start_timer);
    Value::Module(Rc::new(RefCell::new(ns)))
}

fn insert_fn(
    ns: &mut HashMap<String, Value>,
    name: &str,
    func: fn(&[Value]) -> Result<Value, String>,
) {
    ns.insert(
        name.to_string(),
        Value::NativeFunction {
            name: format!("time::{}", name),
            func,
        },
    );
}

fn timer_method(name: &str, func: fn(&[Value]) -> Result<Value, String>) -> Value {
    Value::NativeFunction {
        name: format!("Timer::{}", name),
        func,
    }
}

fn timer_id_from_self(value: &Value) -> Result<i64, String> {
    let s = match value {
        Value::Struct(s) => s,
        _ => return Err("Timer method expects Timer object".to_string()),
    };

    let guard = s.borrow();
    match guard.get("__timer_id") {
        Some(Value::Int(id)) => Ok(*id),
        _ => Err("Invalid Timer object".to_string()),
    }
}

fn build_timer() -> Value {
    let id = TIMER_SEQ.fetch_add(1, Ordering::Relaxed);
    if let Ok(mut map) = timers().lock() {
        map.insert(id, Instant::now());
    }

    let mut fields = HashMap::new();
    fields.insert("__timer_id".to_string(), Value::Int(id));
    fields.insert(
        "elapsedMs".to_string(),
        timer_method("elapsedMs", timer_elapsed_ms),
    );
    fields.insert(
        "elapsedSec".to_string(),
        timer_method("elapsedSec", timer_elapsed_sec),
    );

    Value::Struct(Rc::new(RefCell::new(fields)))
}

fn elapsed(id: i64) -> Result<Duration, String> {
    let map = timers()
        .lock()
        .map_err(|_| "Timer store lock poisoned".to_string())?;
    let start = map
        .get(&id)
        .ok_or_else(|| "Timer has expired or is invalid".to_string())?;
    Ok(start.elapsed())
}

pub fn time_now(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("time.now expects no arguments".to_string());
    }

    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|e| format!("time.now failed: {}", e))?
        .as_millis() as i64;
    Ok(Value::Int(millis))
}

pub fn time_sleep(args: &[Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err("time.sleep expects 1 argument (ms)".to_string());
    }

    let ms = args[1].to_int()?;
    if ms < 0 {
        return Err("time.sleep expects non-negative milliseconds".to_string());
    }
    thread::sleep(Duration::from_millis(ms as u64));
    Ok(Value::Null)
}

pub fn time_hrtime_ms(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("time.hrtimeMs expects no arguments".to_string());
    }
    let boot = BOOT_INSTANT.get_or_init(Instant::now);
    Ok(Value::Float(boot.elapsed().as_secs_f64() * 1000.0))
}

pub fn time_start_timer(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("time.startTimer expects no arguments".to_string());
    }
    Ok(build_timer())
}

fn timer_elapsed_ms(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("timer.elapsedMs expects no arguments".to_string());
    }
    let id = timer_id_from_self(&args[0])?;
    let ms = elapsed(id)?.as_secs_f64() * 1000.0;
    Ok(Value::Float(ms))
}

fn timer_elapsed_sec(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("timer.elapsedSec expects no arguments".to_string());
    }
    let id = timer_id_from_self(&args[0])?;
    let sec = elapsed(id)?.as_secs_f64();
    Ok(Value::Float(sec))
}

File: CRumina/src/builtin/time.cc
#include <builtin/time.h>

#include <chrono>
#include <thread>
#include <atomic>
#include <mutex>
#include <unordered_map>

namespace rumina {
namespace builtin {

static std::atomic<int64_t> timer_seq{1};
static std::mutex timers_mutex;
static std::unordered_map<int64_t, std::chrono::steady_clock::time_point> timers;
static std::chrono::steady_clock::time_point boot_time = std::chrono::steady_clock::now();

// 创建时间模块
Value create_time_module() {
    auto ns = std::make_shared<std::unordered_map<std::string, Value>>();
    
    (*ns)["now"] = Value::makeNativeFunction("time::now", time_now);
    (*ns)["hrtimeMs"] = Value::makeNativeFunction("time::hrtimeMs", time_hrtime_ms);
    (*ns)["sleep"] = Value::makeNativeFunction("time::sleep", time_sleep);
    (*ns)["startTimer"] = Value::makeNativeFunction("time::startTimer", time_start_timer);
    
    return Value::makeModule(ns);
}

// 构建Timer对象
static Value build_timer(int64_t id) {
    auto fields = std::make_shared<std::unordered_map<std::string, Value>>();
    
    (*fields)["__timer_id"] = Value(id);
    (*fields)["elapsedMs"] = Value::makeNativeFunction("Timer::elapsedMs", timer_elapsed_ms);
    (*fields)["elapsedSec"] = Value::makeNativeFunction("Timer::elapsedSec", timer_elapsed_sec);
    
    return Value::makeStruct(fields);
}

// 从Timer对象获取ID
static int64_t timer_id_from_self(const Value& value) {
    if (value.getType() != Value::Type::Struct) {
        throw std::runtime_error("Timer method expects Timer object");
    }
    
    auto fields = value.getStruct();
    auto it = fields->find("__timer_id");
    if (it == fields->end() || it->second.getType() != Value::Type::Int) {
        throw std::runtime_error("Invalid Timer object");
    }
    
    return it->second.getInt();
}

// 获取经过的时间
static std::chrono::duration<double> elapsed(int64_t id) {
    std::lock_guard<std::mutex> lock(timers_mutex);
    auto it = timers.find(id);
    if (it == timers.end()) {
        throw std::runtime_error("Timer has expired or is invalid");
    }
    return std::chrono::steady_clock::now() - it->second;
}

// time.now()
Value time_now(const std::vector<Value>& args) {
    if (!args.empty()) {
        throw std::runtime_error("time.now expects no arguments");
    }
    
    auto now = std::chrono::system_clock::now();
    auto duration = now.time_since_epoch();
    auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
    
    return Value(static_cast<int64_t>(millis));
}

// time.hrtimeMs()
Value time_hrtime_ms(const std::vector<Value>& args) {
    if (!args.empty()) {
        throw std::runtime_error("time.hrtimeMs expects no arguments");
    }
    
    auto duration = std::chrono::steady_clock::now() - boot_time;
    double ms = std::chrono::duration<double, std::milli>(duration).count();
    return Value(ms);
}

// time.sleep(ms)
Value time_sleep(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("time.sleep expects 1 argument (ms)");
    }
    
    int64_t ms = args[0].toInt();
    if (ms < 0) {
        throw std::runtime_error("time.sleep expects non-negative milliseconds");
    }
    
    std::this_thread::sleep_for(std::chrono::milliseconds(ms));
    return Value();
}

// time.startTimer()
Value time_start_timer(const std::vector<Value>& args) {
    if (!args.empty()) {
        throw std::runtime_error("time.startTimer expects no arguments");
    }
    
    int64_t id = timer_seq.fetch_add(1);
    
    {
        std::lock_guard<std::mutex> lock(timers_mutex);
        timers[id] = std::chrono::steady_clock::now();
    }
    
    return build_timer(id);
}

// timer.elapsedMs()
Value timer_elapsed_ms(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("timer.elapsedMs expects no arguments");
    }
    
    int64_t id = timer_id_from_self(args[0]);
    double ms = elapsed(id).count() * 1000.0;
    return Value(ms);
}

// timer.elapsedSec()
Value timer_elapsed_sec(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("timer.elapsedSec expects no arguments");
    }
    
    int64_t id = timer_id_from_self(args[0]);
    double sec = elapsed(id).count();
    return Value(sec);
}

} // namespace builtin
} // namespace rumina
